plugins/node/opentelemetry-instrumentation-express/src/instrumentation.ts

```diff
@@ -313,6 +310,25 @@ export class ExpressInstrumentation extends InstrumentationBase {
           }
         }
       };
+
+      // `handle` isn't just a regular function in some cases. It also contains
+      // some properties holding metadata and state so we need to proxy them
+      // through through patched function
+      // ref: https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1950
+      Object.keys(original).forEach(key => {
+        Object.defineProperty(patched, key, {
+          get() {
+            // @ts-expect-error -- the original function has hidden props indexed by strings
```

Comment:
While I understand the purpose of what we're trying to achieve, the error that comes up here feels very relevant and makes me wonder if there's a better way to do this.

Reply:
@JamieDanielson 

Thanks for your review :) 

> ...makes me wonder if there's a better way to do this.

For "better" you meant at the type level or you think the usage of `Object.defineProperty` is not a good solution? I'll put my thoughts on both.

**Type error**
The type error we get is:
```
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'Function'.
  No index signature with a parameter of type 'string' was found on type 'Function'.ts(7053)
```
The internal typings define the `handle` property of the layer as a Function. I've also checked `@types/express` to see if we can get the type from there but the router stack (where the layers are stored) is defines as `any[]`. So at the time of writing it adding comment to highlight express is "leveraging" the Object nature of functions to hide metadata felt the right thing to do. Note: the type error also arises with @blumamir's proposal since its a type level error.

We may get rid of the error by doing some improvements in the internal types so the compiler will know the function being patched is also used as a record. Something like the type below should work.
```ts
type HandlerType = Function & Record <string,any>
```

**Object.defineProperty vs copy**

Doing the shallow copy proposed by @blumamir fixes this issue as well but those properties represents the state of the layer. If that state is modified for the `handler` to change it's behaviour the original one will not notice that change and keep behaving according the original state (before the copy). With `defineProperty` we keep patched and original in sync (sharing state) so the instrumentation will not interfere with any code accessing or changing the state.

Reply:
Ooh, that sounds like a good idea to use that new `HandlerType`. Would you be open to adding that into `internal-types`, and using that instead of ignoring the TypeScript error?

Also thank you for the detailed explanation, that helps me understand much better!

Reply:
> Ooh, that sounds like a good idea to use that new `HandlerType`. Would you be open to adding that into `internal-types`, and using that instead of ignoring the TypeScript error?
> 
> Also thank you for the detailed explanation, that helps me understand much better!

I support doing this

Reply:
I'll update the types :)

Reply:
Done @JamieDanielson :)